src/mixins/eraser_brush.mixin.js

```diff
@@ -357,192 +326,95 @@
+       * If brush is inverted render all erasable objects with their clip path inverted.
+       * This will render the erased parts as if they were not erased.
        *
        * @param {fabric.Collection} collection
        */
-      prepareCollectionTraversal: function (collection) {
-        var _this = this;
+      _prepareCollectionTraversal: function (collection, ctx) {
         collection.forEachObject(function (obj) {
           if (obj.forEachObject && obj.erasable === 'deep') {
-            _this.prepareCollectionTraversal(obj);
-          }
-          else if (obj.erasable) {
-            _this.hideObject(obj);
+            this._prepareCollectionTraversal(obj, ctx);
           }
-        });
-      },
-
-      /**
-       * @private
-       * Used by {@link fabric.EraserBrush#prepareCanvasObjectsForLayer}
-       * to reverse the action of {@link fabric.EraserBrush#prepareCollectionTraversal}
-       *
-       * @param {fabric.Collection} collection
-       */
-      restoreCollectionTraversal: function (collection) {
-        var _this = this;
-        collection.forEachObject(function (obj) {
-          if (obj.forEachObject && obj.erasable === 'deep') {
-            _this.restoreCollectionTraversal(obj);
+          else if (!obj.erasable && !this.inverted) {
+            obj.render(ctx);
           }
-          else {
-            _this.restoreObjectVisibility(obj);
+          else if (obj.erasable && this.inverted) {
+            var eraser = obj.getEraser();
+            if (eraser) {
+              eraser.inverted = true;
+              obj.dirty = true;
+              obj.render(ctx);
+              eraser.inverted = false;
+              obj.dirty = true;
+            }
           }
-        });
-      },
-
-      /**
-       * @private
-       * This is designed to support erasing a group with both erasable and non-erasable objects.
-       *
-       * @param {'bottom' | 'top' | 'overlay'} layer
-       */
-      prepareCanvasObjectsForLayer: function (layer) {
-        if (layer !== 'bottom') { return; }
-        this.prepareCollectionTraversal(this.canvas);
+        }, this);
       },
 
       /**
+       * Prepare the pattern for the erasing brush
+       * This pattern will be drawn on the top context, achieving a visual effect of erasing only erasable objects
+       * @todo decide how overlay color should behave when `inverted === true`, currently draws over it which is undesirable
        * @private
-       * @param {'bottom' | 'top' | 'overlay'} layer
+       * @param {CanvasRenderingContext2D} ctx
        */
-      restoreCanvasObjectsFromLayer: function (layer) {
-        if (layer !== 'bottom') { return; }
-        this.restoreCollectionTraversal(this.canvas);
+      preparePattern: function (ctx) {
+        if (!this._patternCanvas) {
+          this._patternCanvas = fabric.util.createCanvasElement();
+        }
+        var canvas = this._patternCanvas;
+        canvas.width = this.canvas.width;
```

Comment:
@asturur I think this canvas needs retina scaling.
Tried it but couldn't get the correct size.
Can you help?

Reply:
If I understand correctly I should apply retina scaling to the pattern canvas and then scale top context by `1/retinaScaling`.
However this is an edge case fabric doesn't handle well. Events get transformed as well by this transformation, so it sounds bad to do this.
Ideas?

Reply:
look at the text being erased.
text is not erasable so what you see is the pattern drawing over it.
because retina scaling is accounted for it is blurred
https://user-images.githubusercontent.com/34343793/148683724-a60fff60-d40d-4978-8576-89fcac511498.mp4